### ISSUE TITLE
window I created myself cannot display touchbar

### ISSUE THREAD

**Issue by sxmxta** [LouisBrunner/go-touchbar#2]
Hello, I followed the example and used Install instead of Debug to display touchbar I am using self created windows, how do I need to associate them?

**Comment by LouisBrunner** [comment]
Hi @sxmxta,

Can you show me an example on how you are using it?

When using `Install`, the library should pick up the window returned by `[NSApplication sharedApplication].mainWindow`.

**Comment by sxmxta** [comment]
Hi @LouisBrunner ,

Thank you very much for your reply

Let me first describe the scenarios in which it is used

I am writing an open-source framework, It is developed in the Go language based on CEF4Delphi, but I did not find CEF-related functional support for taouchbar
The window creation here uses Lazarus LCL's component library, and the underlying layer also uses Cocoa
In this case, they do not have support for touchbar
I can run the touchbar library example program you wrote.
But when I put it in my project, it doesn't work. I don't know how to relate them together, and I have no knowledge of Object C

[https://github.com/energye/energy](https://github.com/energye/energy) This is the project address I wrote. If possible, could you help debug it?


The following is the example code, as CEF is a multiprocess model, it will start multiple times when executing the program. This SetBrowserInit function is called by the UI thread when creating a window

```go
package main

import (
	"fmt"
	"github.com/energye/energy/v2/cef"
	"github.com/energye/energy/v2/example/macos/touchbar/bar"
	"github.com/energye/energy/v2/pkgs/touchbar"
	"github.com/energye/energy/v2/pkgs/touchbar/barbuilder"
	"github.com/energye/energy/v2/pkgs/touchbar/barutils"
)

func main() {
	cef.GlobalInit(nil, nil)

	//create application
	app := cef.NewApplication()
	app.SetUseMockKeyChain(true)
	cef.BrowserWindow.Config.Url = "https://www.baidu.com"
	cef.BrowserWindow.SetBrowserInit(func(event *cef.BrowserEvent, window cef.IBrowserWindow) {
		tb := touchbar.New(barbuilder.Options{
			EventErrorLogger: func(err error) {
				fmt.Println("EventErrorLogger", err)
			},
		})

		makeUpdater := func(switcher barutils.Switcher) func() {
			return func() {
				err := switcher.Update()
				if err != nil {
					fmt.Printf("could not update: %v\n", err)
				}
			}
		}

		config := barutils.MakeStackableBar(tb, func(switcher barutils.Switcher) []barbuilder.Item {
			update := makeUpdater(switcher)
			return []barbuilder.Item{
				&barbuilder.Label{
					Content: &barbuilder.ContentLabel{
						Text: "Go Touch Bar",
					},
				},
				&barbuilder.SpaceLarge{},
				bar.MakeDemo(update),
				&barbuilder.SpaceSmall{},
				bar.MakeCatalog(switcher, update),
			}
		})

		err := tb.Install(config)
		fmt.Println("install err", err)
		if err != nil {
			panic(err)
		}

		// end
		//err = tb.Uninstall()
		//if err != nil {
		//	panic(err)
		//}
	})
	//run application
	cef.Run(app)
}

```

**Comment by sxmxta** [comment]
Hi @LouisBrunner ,

Hello, I may have found a solution. Need to call setDelegate

```go
  NSApplication* app = [NSApplication sharedApplication];
  NSWindow* window;
  if (mode == kMainWindow) {
    window = app.mainWindow;
    [app setDelegate:controller]; // new add code
  } else if (mode == kDebug) {
    window = makeDebugWindow(app);
  } else {
    result.err = [[NSString stringWithFormat:@"Unknown mode %lu", mode] UTF8String];
    return result;
  }
```

---

My window was created through the Dylib I encapsulated, and window event management is included in it. I don't know if there will be any problems after adding it like this?

In my project, there is also a systray module implemented in a similar way.

**Comment by LouisBrunner** [comment]
Thanks a lot for the context!

> Hello, I may have found a solution. Need to call setDelegate

Does that solve your issue? If you can confirm, I will get the fix in ASAP.

**Comment by sxmxta** [comment]
> Thanks a lot for the context!
> 
> > Hello, I may have found a solution. Need to call setDelegate
> 
> Does that solve your issue? If you can confirm, I will get the fix in ASAP.

At present, it seems that my problem has been solved.

Run the touchbar example directly in my project, debug=false, and use the window I created myself. The result of the example is the same.

---

Additionally, I noticed that some of the touchbar APIs described in Readme have not yet been implemented. Can you help improve touchbar's other APIs if possible?

Because my open-source project Energy is building a cross platform desktop application framework, it does not yet support touchbar on MacOS

---

Thank you very much for providing Golang's Touch Bar, lib

**Comment by LouisBrunner** [comment]
> At present, it seems that my problem has been solved.

I merged the fix in #3 so the library should now work as-is.

> Additionally, I noticed that some of the touchbar APIs described in Readme have not yet been implemented. Can you help improve touchbar's other APIs if possible?

I didn't need all the features for the application I was making so I didn't implement them all. If you give me a prioritized list of which ones are the most important to you, I could try and add them. You are also welcome to do a PR and add them yourself! 😄
